File: appwrite/functions/game-api/src/domain/mapGenerator.ts
import type { HexCoord } from './gameDomain.js';
import { hexKey } from './gameDomain.js';
import type { TileMap, TerrainType } from './terrain.js';
import { isOnOffsetGrid, offsetNeighbor, offsetWithinRange, orthogonalNeighbors, manhattanDistance } from './hexOffset.js';

export const MAP_SIZE = 16;

class SeededRng {
  private state: number;

  constructor(seed: number) {
    this.state = seed;
  }

  next(): number {
    this.state = (this.state * 1664525 + 1013904223) >>> 0;
    return this.state / 0x100000000;
  }

  nextInt(min: number, max: number): number {
    return min + Math.floor(this.next() * (max - min + 1));
  }
}

function isInSpawnZone(hex: HexCoord): boolean {
  return (hex.q <= 2 || hex.q >= MAP_SIZE - 3) && hex.r >= 4 && hex.r <= 12;
}

function isOnGrid(hex: HexCoord): boolean {
  return isOnOffsetGrid(hex.q, hex.r, MAP_SIZE, MAP_SIZE);
}

function clearRect(tiles: TileMap, colMin: number, colMax: number, rowMin: number, rowMax: number): void {
  for (let row = rowMin; row <= rowMax && row < MAP_SIZE; row++) {
    for (let col = colMin; col <= colMax && col < MAP_SIZE; col++) {
      tiles[hexKey({ q: col, r: row })] = 'Plains';
    }
  }
}

function paintBlob(
  tiles: TileMap,
  center: HexCoord,
  terrain: TerrainType,
  radius: number,
  rng: SeededRng,
  density: number,
): void {
  for (const hex of offsetWithinRange(center, radius)) {
    if (!isOnGrid(hex)) continue;
    if (isInSpawnZone(hex)) continue;

    if (hex.q !== center.q || hex.r !== center.r) {
      if (rng.next() >= density) continue;
    }

    const key = hexKey(hex);
    const existing = tiles[key] ?? 'Plains';

    if (terrain === 'ShallowWater' && (existing === 'Mountain' || existing === 'DeepWater')) continue;
    if (
      terrain === 'Forest' &&
      (existing === 'Mountain' || existing === 'DeepWater' || existing === 'ShallowWater')
    ) {
      continue;
    }
    if (terrain === 'Hill' && existing !== 'Plains') continue;

    tiles[key] = terrain;
  }
}

function randomCoord(rng: SeededRng, min: number, max: number): HexCoord {
  return { q: rng.nextInt(min, max), r: rng.nextInt(min, max) };
}

export function generateMap(seed: number): TileMap {
  const rng = new SeededRng(seed);
  const tiles: TileMap = {};

  for (let row = 0; row < MAP_SIZE; row++) {
    for (let col = 0; col < MAP_SIZE; col++) {
      tiles[hexKey({ q: col, r: row })] = 'Plains';
    }
  }

  paintBlob(tiles, { q: 8, r: 8 }, 'Mountain', 2, rng, 0.65);
  paintBlob(tiles, { q: 7, r: 5 }, 'Mountain', 2, rng, 0.55);
  paintBlob(tiles, { q: 9, r: 11 }, 'Mountain', 1, rng, 0.7);

  for (let i = 0; i < 6; i++) {
    paintBlob(tiles, randomCoord(rng, 4, 11), 'DeepWater', rng.nextInt(1, 2), rng, 0.45);
  }

  for (const key of Object.keys(tiles)) {
    if (tiles[key] !== 'DeepWater') continue;
    const [q, r] = key.split(',').map(Number);
    for (let d = 0; d < 4; d++) {
      const n = offsetNeighbor({ q, r }, d);
      if (!isOnGrid(n)) continue;
      const nKey = hexKey(n);
      if (tiles[nKey] === 'Plains' && rng.next() < 0.55) {
        tiles[nKey] = 'ShallowWater';
      }
    }
  }

  for (let i = 0; i < 14; i++) {
    paintBlob(tiles, randomCoord(rng, 0, MAP_SIZE - 1), 'Forest', rng.nextInt(1, 2), rng, 0.5);
  }

  for (let i = 0; i < 10; i++) {
    paintBlob(tiles, randomCoord(rng, 0, MAP_SIZE - 1), 'Hill', 1, rng, 0.55);
  }

  clearRect(tiles, 0, 2, 4, 12);
  clearRect(tiles, MAP_SIZE - 3, MAP_SIZE - 1, 4, 12);
  return tiles;
}

File: src/CombatGame.Domain/Maps/MapGenerator.cs
using CombatGame.Domain.Hex;

namespace CombatGame.Domain.Maps;

public static class MapGenerator
{
    public const int MapSize = 16;

    public static HexGrid Generate(Random rng)
    {
        var grid = new HexGrid(MapSize, MapSize);

        foreach (var tile in grid.Tiles.Values)
        {
            tile.Terrain = TerrainType.Plains;
        }

        PaintBlob(grid, HexOffset.FromOddR(8, 8), TerrainType.Mountain, 2, rng, 0.65);
        PaintBlob(grid, HexOffset.FromOddR(7, 5), TerrainType.Mountain, 2, rng, 0.55);
        PaintBlob(grid, HexOffset.FromOddR(9, 11), TerrainType.Mountain, 1, rng, 0.7);

        for (var i = 0; i < 6; i++)
        {
            var center = RandomCoord(rng, 4, 11);
            PaintBlob(grid, center, TerrainType.DeepWater, rng.Next(1, 3), rng, 0.45);
        }

        foreach (var tile in grid.Tiles.Values)
        {
            if (tile.Terrain != TerrainType.DeepWater)
            {
                continue;
            }

            for (var d = 0; d < 6; d++)
            {
                var neighbor = tile.Coord.Neighbor(d);
                if (!grid.Contains(neighbor))
                {
                    continue;
                }

                if (grid.GetTerrain(neighbor) == TerrainType.Plains && rng.NextDouble() < 0.55)
                {
                    grid.SetTerrain(neighbor, TerrainType.ShallowWater);
                }
            }
        }

        for (var i = 0; i < 14; i++)
        {
            PaintBlob(grid, RandomCoord(rng, 0, MapSize - 1), TerrainType.Forest, rng.Next(1, 3), rng, 0.5);
        }

        for (var i = 0; i < 10; i++)
        {
            PaintBlob(grid, RandomCoord(rng, 0, MapSize - 1), TerrainType.Hill, rng.Next(1, 2), rng, 0.55);
        }

        EnsureSpawnZones(grid);
        grid.EnsureAllTiles();
        return grid;
    }

    private static HexCoord RandomCoord(Random rng, int min, int max)
    {
        var col = rng.Next(min, max + 1);
        var row = rng.Next(min, max + 1);
        return HexOffset.FromOddR(col, row);
    }

    private static void EnsureSpawnZones(HexGrid grid)
    {
        ClearRect(grid, 0, 2, 4, 12);
        ClearRect(grid, MapSize - 3, MapSize - 1, 4, 12);
    }

    private static void ClearRect(HexGrid grid, int colMin, int colMax, int rowMin, int rowMax)
    {
        for (var row = rowMin; row <= rowMax && row < grid.Height; row++)
        {
            for (var col = colMin; col <= colMax && col < grid.Width; col++)
            {
                grid.SetTerrain(HexOffset.FromOddR(col, row), TerrainType.Plains);
            }
        }
    }

    private static void PaintBlob(
        HexGrid grid,
        HexCoord center,
        TerrainType terrain,
        int radius,
        Random rng,
        double density)
    {
        foreach (var hex in center.WithinRange(radius))
        {
            if (!grid.Contains(hex))
            {
                continue;
            }

            if (IsInSpawnZone(hex))
            {
                continue;
            }

            if (hex.DistanceTo(center) == 0 || rng.NextDouble() < density)
            {
                var existing = grid.GetTerrain(hex);
                if (terrain == TerrainType.ShallowWater && existing is TerrainType.Mountain or TerrainType.DeepWater)
                {
                    continue;
                }

                if (terrain == TerrainType.Forest && existing is TerrainType.Mountain or TerrainType.DeepWater or TerrainType.ShallowWater)
                {
                    continue;
                }

                if (terrain == TerrainType.Hill && existing is not TerrainType.Plains)
                {
                    continue;
                }

                grid.SetTerrain(hex, terrain);
            }
        }
    }

    private static bool IsInSpawnZone(HexCoord coord)
    {
        var (col, row) = HexOffset.ToOddR(coord);
        return (col <= 2 || col >= MapSize - 3) && row is >= 4 and <= 12;
    }
}
